File: Data Structures & Algorithms/count-paths/submission-0.java
class Solution {
    public int uniquePaths(int m, int n) {
        int r = 0, c= 0;
        Integer[][] dp = new Integer[m][n];
        return helper (r, c, m, n, dp);
    }

    int helper (int r, int c, int m, int n, Integer[][] dp) {
        if (r == m-1 || c == n-1)
            return 1;
        if (dp[r][c] != null) 
            return dp[r][c];

        int right = helper (r, c+1, m, n, dp);
        int down = helper (r+1, c, m, n, dp);

        return dp[r][c] = right + down;
    }
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        map_ = {val: i for i, val in enumerate(inorder)}
        self.indx = -1

        def helper(start, end):
            if start > end:
                return None

            self.indx += 1
            root = TreeNode(preorder[self.indx])
            
            inorder_indx = map_[root.val]
            
            root.left = helper(start, inorder_indx - 1)
            root.right = helper(inorder_indx + 1, end)

            return root

        return helper(0, len(preorder) - 1)

